File: ui/src/api/hooks/recipes.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { commands, type AgentRecipe, type AgentRecipeRun } from "../client";

type AppLikeError = Error & { code?: string };

const recipesKey = ["recipes"] as const;

function toError(message: string, code?: string): never {
  const err = new Error(message) as AppLikeError;
  err.code = code;
  throw err;
}

export function useRecipes(includePaused = false) {
  return useQuery<AgentRecipe[]>({
    queryKey: [...recipesKey, includePaused],
    queryFn: async () => {
      const result = await commands.listRecipes(includePaused);
      if (result.status === "error") toError(result.error.message, result.error.code);
      return result.data;
    },
    staleTime: 30_000,
  });
}

export function useCreateRecipe() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (input: {
      title: string;
      description: string;
      recipeKind: string;
      promptTemplate: string;
      cadence: string;
      dayOfWeek: number | null;
      dayOfMonth: number | null;
    }) => {
      const result = await commands.createRecipe(
        input.title,
        input.description,
        input.recipeKind,
        input.promptTemplate,
        input.cadence,
        input.dayOfWeek,
        input.dayOfMonth,
      );
      if (result.status === "error") toError(result.error.message, result.error.code);
      return result.data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: recipesKey });
    },
  });
}

export function useUpdateRecipe() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (input: {
      id: string;
      title: string;
      description: string;
      promptTemplate: string;
      cadence: string;
      dayOfWeek: number | null;
      dayOfMonth: number | null;
    }) => {
      const result = await commands.updateRecipe(
        input.id,
        input.title,
        input.description,
        input.promptTemplate,
        input.cadence,
        input.dayOfWeek,
        input.dayOfMonth,
      );
      if (result.status === "error") toError(result.error.message, result.error.code);
      return result.data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: recipesKey });
    },
  });
}

export function usePauseRecipe() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      const result = await commands.pauseRecipe(id);
      if (result.status === "error") toError(result.error.message, result.error.code);
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: recipesKey });
    },
  });
}

export function useResumeRecipe() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      const result = await commands.resumeRecipe(id);
      if (result.status === "error") toError(result.error.message, result.error.code);
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: recipesKey });
    },
  });
}

export function useDeleteRecipe() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      const result = await commands.deleteRecipe(id);
      if (result.status === "error") toError(result.error.message, result.error.code);
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: recipesKey });
    },
  });
}

export function useTriggerRecipe() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      const result = await commands.triggerRecipe(id);
      if (result.status === "error") toError(result.error.message, result.error.code);
      return result.data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: recipesKey });
      void qc.invalidateQueries({ queryKey: ["action-bundles"] });
      void qc.invalidateQueries({ queryKey: ["recipe-runs"] });
    },
  });
}

export function useRecipeRuns(recipeId: string) {
  return useQuery<AgentRecipeRun[]>({
    queryKey: ["recipe-runs", recipeId],
    queryFn: async () => {
      if (!recipeId) return [];
      const result = await commands.listRecipeRuns(recipeId, null);
      if (result.status === "error") toError(result.error.message, result.error.code);
      return result.data;
    },
    enabled: recipeId.length > 0,
    staleTime: 30_000,
  });
}

File: crates/finsight-app/src/commands/recipes.rs
use crate::error::AppResult;
use crate::AppState;
use finsight_core::models::{AgentRecipe, AgentRecipeRun};
use tauri::State;

#[tauri::command]
#[specta::specta]
pub async fn list_recipes(
    state: State<'_, AppState>,
    include_paused: bool,
) -> AppResult<Vec<AgentRecipe>> {
    finsight_api::commands::recipes::list_recipes(&state.api, include_paused).await
}

#[tauri::command]
#[specta::specta]
pub async fn create_recipe(
    state: State<'_, AppState>,
    title: String,
    description: String,
    recipe_kind: String,
    prompt_template: String,
    cadence: String,
    day_of_week: Option<i64>,
    day_of_month: Option<i64>,
) -> AppResult<AgentRecipe> {
    finsight_api::commands::recipes::create_recipe(
        &state.api,
        title,
        description,
        recipe_kind,
        prompt_template,
        cadence,
        day_of_week,
        day_of_month,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn update_recipe(
    state: State<'_, AppState>,
    id: String,
    title: String,
    description: String,
    prompt_template: String,
    cadence: String,
    day_of_week: Option<i64>,
    day_of_month: Option<i64>,
) -> AppResult<AgentRecipe> {
    finsight_api::commands::recipes::update_recipe(
        &state.api,
        id,
        title,
        description,
        prompt_template,
        cadence,
        day_of_week,
        day_of_month,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn pause_recipe(state: State<'_, AppState>, id: String) -> AppResult<()> {
    finsight_api::commands::recipes::pause_recipe(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn resume_recipe(state: State<'_, AppState>, id: String) -> AppResult<()> {
    finsight_api::commands::recipes::resume_recipe(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_recipe(state: State<'_, AppState>, id: String) -> AppResult<()> {
    finsight_api::commands::recipes::delete_recipe(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn trigger_recipe(state: State<'_, AppState>, id: String) -> AppResult<String> {
    finsight_api::commands::recipes::trigger_recipe(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_recipe_runs(
    state: State<'_, AppState>,
    recipe_id: String,
    limit: Option<u32>,
) -> AppResult<Vec<AgentRecipeRun>> {
    finsight_api::commands::recipes::list_recipe_runs(&state.api, recipe_id, limit).await
}
